components/webext-storage/src/sync/mod.rs

```diff
@@ -0,0 +1,156 @@
+// Main problem is that it doesn't handle deletions - but to do that, we need
+// something other than a simple Option<JsonMap> - we need to differentiate
+// "doesn't exist" from "removed".
+// TODO!
+fn merge(other: JsonMap, mut ours: JsonMap, parent: Option<JsonMap>) -> IncomingAction {
+    if other == ours {
+        return IncomingAction::Same;
+    }
+    // Server wins. Iterate over incoming - if incoming and the parent are
+    // identical, then we will take our local value.
+    for (k, iv) in other.into_iter() {
+        let vour = ours.get(&k);
+        match vour {
+            Some(vour) => {
+                if *vour != iv {
+                    // So we have a discrepency between 'ours' and 'other' - use parent
+                    // to resolve.
+                    let can_take_local = match parent {
+                        Some(ref pm) => {
+                            if let Some(pv) = pm.get(&k) {
+                                // parent has a value - we can only take our local
+                                // value if the parent and incoming have the same.
+                                *pv == iv
+                            } else {
+                                // Value doesn't exist in the parent - can't take local
+                                false
+                            }
+                        }
+                        None => {
+                            // 2 way merge because there's no parent. We always
+                            // prefer incoming here.
+                            false
+                        }
+                    };
+                    if can_take_local {
+                        log::trace!("merge: no remote change in key {} - taking local", k);
+                    } else {
+                        log::trace!("merge: conflict in existing key {} - taking remote", k);
+                        ours.insert(k, iv);
+                    }
+                } else {
+                    log::trace!("merge: local and incoming same for key {}", k);
+                }
+            }
+            None => {
+                log::trace!("merge: incoming new value for key {}", k);
+                ours.insert(k, iv);
+            }
+        }
+    }
+    IncomingAction::Merge { data: ours }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::error::*;
+    use serde_json::json;
+
+    // a macro for these tests - constructs a serde_json::Value::Object
+    macro_rules! map {
+        ($($map:tt)+) => {
+            json!($($map)+).as_object().unwrap().clone()
+        };
+    }
+
+    #[test]
+    fn test_3way_merging() -> Result<()> {
+        // No conflict - identical local and remote.
+        assert_eq!(
+            merge(
+                map!({"one": "one", "two": "two"}),
+                map!({"two": "two", "one": "one"}),
+                Some(map!({"parent_only": "parent"})),
```

Comment:
Hmm, in this case, we still want to update the mirror without reuploading, though, no? (Or maybe it's fine to leave it as is, if it can't cause problems for future merges...say, if local or remote ever brings `parent_only` back).

Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

Reply:
> Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

Yep - the final outgoing process writes staging to the mirror.